File: avr-mdb/crc_test.c
#include <inttypes.h>
#include <stdio.h>
#include "crc.h"

void test1(uint8_t const b) {
  printf("b1=%02x crc=%02x\n", b, crc8_p93_next(0, b));
}

void test2(uint8_t const b1, uint8_t const b2) {
  printf("b2=%02x%02x crc=%02x\n", b1, b2, crc8_p93_2b(b1, b2));
}

void test_str(uint8_t const bs[], uint8_t const len) {
  uint8_t crc = 0;
  printf("bs=");
  for (uint8_t i = 0; i < len; i++) {
    printf("%02x", bs[i]);
    crc = crc8_p93_next(crc, bs[i]);
  }
  printf(" crc=%02x\n", crc);
}

int main() {
  test1(0x00);
  test1(0x55);
  test1(0xaa);
  test1(0xff);

  test2(0x80, 0x00);
  test2(0x30, 0x30);
  test2(0x20, 0x30);
  test2(0xe0, 0x69);
  test2(0xe0, 0x78);

  test2(0x03, 0x01);
  test_str((uint8_t[]){0x06, 0x00, 0xbe, 0xeb, 0xee}, 5);
  test_str((uint8_t[]){0x04, 0x0f, 0x30}, 3);
  test2(0x03, 0x00);
  test2(0x03, 0x08);
  return 0;
}

File: crc/crc_test.go
package crc

import (
	"strings"
	"testing"
)

func makeCheck2(fun func(byte, byte) byte, tag string) func(t *testing.T, v1, v2, expect byte) {
	return func(t *testing.T, v1, v2, expect byte) {
		if fun(v1, v2) != expect {
			t.Errorf("%s(%02x, %02x) != %02x", tag, v1, v2, expect)
		}
	}
}

func makeCheckN(fun func(byte, []byte) byte, tag string) func(t *testing.T, v1 byte, vs []byte, expect byte) {
	return func(t *testing.T, v1 byte, vs []byte, expect byte) {
		if fun(v1, vs) != expect {
			t.Errorf("%s(%02x, "+strings.Repeat("%02x", len(vs))+") != %02x", tag, v1, vs, expect)
		}
	}
}

func TestReference(t *testing.T) {
	checkRef := makeCheck2(CRC8_p93_reference, "CRC8_p93_reference")
	checkRef(t, 0, 0x00, 0x00)
	checkRef(t, 0, 0x55, 0x86)
	checkRef(t, 0, 0xaa, 0x9f)
	checkRef(t, 0, 0xff, 0x19)
}

func TestLookup(t *testing.T) {
	checkNext := makeCheck2(CRC8_p93_next, "CRC8_p93_next")
	checkNext(t, 0, 0x00, 0x00)
	checkNext(t, 0, 0x55, 0x86)
	checkNext(t, 0, 0xaa, 0x9f)
	checkNext(t, 0, 0xff, 0x19)
	check2 := makeCheck2(CRC8_p93_2, "CRC8_p93_2")
	check2(t, 0x80, 0x00, 0x74)
	check2(t, 0xe0, 0x78, 0xc9)
	check2(t, 0x03, 0x01, 0xc8)
	checkN := makeCheckN(CRC8_p93_n, "CRC8_p93_n")
	checkN(t, 0, []byte{0x06, 0x00, 0xbe, 0xeb, 0xee}, 0x75)
	checkN(t, 0, []byte{0x04, 0x0f, 0x30}, 0xf7)
}
